### PR TITLE
Populate HandleInput

### LINKED ISSUE #1235: Populate HandleInput based on existence of overridden input-handling methods
As a follow-up from https://github.com/ppy/osu-framework/issues/952, we still want to populate `HandleInput` automatically to further nail down drawables that are unnecessarily checked for input.

The method to do this is explained here: https://github.com/ppy/osu-framework/pull/1135#issuecomment-344797281. Heavy caching will be required to not result in a performance loss.

### PR DESCRIPTION
**Pull request by UselessToucan**
resolves #1235

### CODE REVIEW

**Comment by UselessToucan** [comment]
It looks like I can't add reviewers, but I still should summon @default0 here somehow.

File: osu.Framework/Caching/HandleInputCache.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+using osu.Framework.Graphics;
+
+namespace osu.Framework.Caching
+{
+    public class HandleInputCache
+    {
+        private readonly ConcurrentDictionary<Type, bool> cachedValues = new ConcurrentDictionary<Type, bool>();
+
+        private readonly string[] inputMethods = {
+            "OnHover",
```
Comment (peppy) [inline osu.Framework/Caching/HandleInputCache.cs]: can't use `nameof` for these?

File: osu.Framework/Caching/HandleInputCache.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu-framework/master/LICENCE
+
+using System;
+using System.Collections.Concurrent;
+using System.Reflection;
+using osu.Framework.Graphics;
+
+namespace osu.Framework.Caching
+{
+    public class HandleInputCache
```
Comment (default0) [inline osu.Framework/Caching/HandleInputCache.cs]: Any reason to expose this class publicly? Shouldn't this cache be internal? Or is there any use Applications can make of it? Also, any reason to have this cache be non-static? Other than "oh I can use DI with it" (which should be unnecessary - as should be wasting a field on holding an instance of this in each Drawable).

File: osu.Framework/Graphics/Drawable.cs [hunk 3]
```diff
@@ -1830,10 +1836,31 @@ protected virtual void OnFocusLost(InputState state)
         /// is propagated up the scene graph to the next eligible Drawable.</returns>
         protected virtual bool OnMouseMove(InputState state) => false;
 
+        private HandleInputCache handleInputCache;
+
+        public string[] InputMethods => new[]
```
Comment (default0) [inline osu.Framework/Graphics/Drawable.cs]: Again, any reason to have this public?

File: osu.Framework/Caching/HandleInputCache.cs (hunk 2 above)
Comment (UselessToucan) [reply osu.Framework/Caching/HandleInputCache.cs]: No reasons. I just did it this way.

**Comment by peppy** [comment]
I like this because it takes out the manual overhead of remembering to add `HandleInput`, but still allows the override to exist for dynamic scenarios.

**Comment by peppy** [comment]
Let's adjust the `HandleInput` xmldoc before merging this.

**Comment by smoogipoo** [comment]
@peppy plz approve.

**Pull request merged**